diff -r 18fd959b07ef -r d9f82c5cf1f5 src/ex_cmds.h
--- a/src/ex_cmds.h	Wed Aug 13 22:05:54 2014 +0200
+++ b/src/ex_cmds.h	Tue Aug 12 10:31:37 2014 +0100
@@ -110,7 +110,7 @@
 EX(CMD_args,		"args",		ex_args,
 			BANG|FILES|EDITCMD|ARGOPT|TRLBAR),
 EX(CMD_argadd,		"argadd",	ex_argadd,
-			BANG|NEEDARG|RANGE|NOTADR|ZEROR|FILES|TRLBAR),
+			BANG|RANGE|NOTADR|ZEROR|FILES|TRLBAR),
 EX(CMD_argdelete,	"argdelete",	ex_argdelete,
 			BANG|RANGE|NOTADR|FILES|TRLBAR),
 EX(CMD_argdo,		"argdo",	ex_listdo,
diff -r 18fd959b07ef -r d9f82c5cf1f5 src/ex_cmds2.c
--- a/src/ex_cmds2.c	Wed Aug 13 22:05:54 2014 +0200
+++ b/src/ex_cmds2.c	Tue Aug 12 10:31:37 2014 +0100
@@ -1946,6 +1946,11 @@
 #endif
 
     /*
+     * Set default argument for ":argadd" command.
+     */
+    if (what == AL_ADD && STRLEN(str) == 0)
+	str = curbuf->b_fname;
+    /*
      * Collect all file name arguments in "new_ga".
      */
     if (get_arglist(&new_ga, str) == FAIL)
